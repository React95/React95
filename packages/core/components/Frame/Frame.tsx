import React, {
  ComponentPropsWithRef,
  ElementType,
  ForwardedRef,
  forwardRef,
} from 'react';
import { Sprinkles, sprinkles } from './sprinkles.css';
import cn from 'classnames';

type FixedForwardRef = <T, P = object>(
  render: (props: P, ref: React.Ref<T>) => React.ReactNode,
) => (props: P & React.RefAttributes<T>) => React.ReactNode;

const fixedForwardRef = forwardRef as FixedForwardRef;

type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

export type FrameProps<TAs extends ElementType = 'div'> = {
  as?: TAs;
} & DistributiveOmit<
  ComponentPropsWithRef<ElementType extends TAs ? 'div' : TAs>,
  'as'
> &
  Sprinkles;

const FrameComponent = <TAs extends ElementType>(
  props: FrameProps<TAs>,
  ref: ForwardedRef<any>,
) => {
  const { as, children, ...rest } = props;
  const Component = as || 'div';
  const { className, style, otherProps } = sprinkles(rest);

  return (
    <Component
      {...otherProps}
      style={{ ...style, ...otherProps.style }}
      className={cn(className, otherProps.className)}
      ref={ref}
    >
      {children}
    </Component>
  );
};

export const Frame = fixedForwardRef(FrameComponent);
