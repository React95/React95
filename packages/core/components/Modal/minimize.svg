<svg width='6' height='2' fill='currentColor' xmlns='http://www.w3.org/2000/svg'><path d='M0 0h6v2H0z'/></svg>