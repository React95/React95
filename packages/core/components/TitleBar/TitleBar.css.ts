import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';
import { recipe } from '@vanilla-extract/recipes';

export const optionsBox = style({
  listStyle: 'none',
  padding: 0,
  margin: 0,
  display: 'flex',
});

export const option = style({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  marginLeft: contract.space[2],
  padding: contract.space[0],
  width: contract.space[17],
  height: contract.space[14],
  minWidth: contract.space[0],
  fontWeight: 600,
  fontSize: 10,
  selectors: {
    '&:active': {
      paddingTop: contract.space[1],
      paddingRight: contract.space[0],
      paddingBottom: contract.space[0],
      paddingLeft: contract.space[1],
      outline: 'none',
    },
    '&:first-child': {
      marginRight: contract.space[0],
    },
    '&:focus': {
      boxShadow: `inset 1px 1px 0px 1px ${contract.colors.borderLightest},
        inset -1px -1px 0px 1px ${contract.colors.borderDark}`,
    },
  },
});

export const titleBarBackground = recipe({
  base: {
    height: contract.space[20],
    marginBottom: contract.space[2],
    padding: contract.space[2],
    display: 'flex',
    boxShadow: 'none',
    userSelect: 'none',
  },
  variants: {
    active: {
      true: {
        background: contract.colors.headerBackground,
        color: contract.colors.headerText,
      },
      false: {
        background: contract.colors.headerNotActiveBackground,
        color: contract.colors.headerNotActiveText,
      },
    },
  },
});

globalStyle(`${titleBarBackground.classNames.base} svg`, {
  width: contract.space[16],
  height: contract.space[16],
  marginRight: contract.space[4],
});

export const title = style({
  flexGrow: 1,
  lineHeight: '1.4em',
  margin: '0',
  fontSize: '1em',
  textShadow: '0.5px 0px white, 1.5px 0px white',
  color: 'transparent',
  letterSpacing: '1px',
});
