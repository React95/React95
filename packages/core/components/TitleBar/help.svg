<svg width='6' height='9' fill='currentColor' xmlns='http://www.w3.org/2000/svg'><path d='M0 1h2v2H0zM1 0h4v1H1zM4 1h2v2H4zM3 3h2v1H3zM2 4h2v2H2zM2 7h2v2H2z'/></svg>