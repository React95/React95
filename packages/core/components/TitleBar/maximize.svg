<svg width='9' height='9' fill='currentColor' xmlns='http://www.w3.org/2000/svg'><path fill-rule='evenodd' clip-rule='evenodd' d='M9 0H0v9h9V0zM8 2H1v6h7V2z' /></svg>