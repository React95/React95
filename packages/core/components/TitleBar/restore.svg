<svg width='8' height='9' fill='currentColor' xmlns='http://www.w3.org/2000/svg'><path d='M2 0h6v2H2zM7 2h1v4H7zM2 2h1v1H2zM6 5h1v1H6zM0 3h6v2H0zM5 5h1v4H5zM0 5h1v4H0zM1 8h4v1H1z'/></svg>