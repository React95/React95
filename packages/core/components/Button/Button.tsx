import React, { forwardRef } from 'react';

import { button } from './Button.css';
import Frame from '../Frame';
import { FrameProps } from '../Frame/Frame';
import cn from 'classnames';

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  FrameProps;

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    return (
      <Frame
        as="button"
        {...props}
        className={cn(button, props.className)}
        ref={ref}
      />
    );
  },
);
