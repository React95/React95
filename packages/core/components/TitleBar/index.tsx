import TitleBar from './TitleBar';

export default TitleBar;
