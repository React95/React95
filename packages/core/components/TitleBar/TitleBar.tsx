import cn from 'classnames';
import type {
  ButtonHTMLAttributes,
  ElementRef,
  ElementType,
  ForwardedRef,
  ForwardRefRenderFunction,
  HTMLAttributes,
  ReactElement,
  Ref,
} from 'react';
import React from 'react';

import close from './close.svg';
import help from './help.svg';
import maximize from './maximize.svg';
import minimize from './minimize.svg';
import restore from './restore.svg';

import { fixedForwardRef, Frame, FrameProps } from '../Frame/Frame';

import { button } from '../Button/Button.css';
import * as styles from './TitleBar.css';

const defaultOptions = {
  help: {
    alt: 'help',
    src: help,
  },
  close: {
    alt: 'close',
    src: close,
  },
  maximize: {
    alt: 'maximize',
    src: maximize,
  },
  minimize: {
    alt: 'minimize',
    src: minimize,
  },
  restore: {
    alt: 'restore',
    src: restore,
  },
};

const OptionsBox = fixedForwardRef<HTMLDivElement, FrameProps<'div'>>(
  (rest, ref) => (
    <Frame
      {...rest}
      ref={ref}
      className={cn(styles.optionsBox, rest.className)}
    />
  ),
);

export type OptionProps<TAs extends ElementType> =
  ButtonHTMLAttributes<HTMLButtonElement> & FrameProps<TAs>;

export type OptionReturnType = <TAs extends ElementType = 'button'>(
  props: OptionProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;

const Option = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (rest, ref) => (
    <Frame
      as="button"
      {...rest}
      ref={ref}
      className={cn(button, styles.option, rest.className)}
    />
  ),
) as <TAs extends ElementType = 'button'>(
  props: OptionProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;

type DefaultOptionProps<TAs extends ElementType = 'button'> = {
  kind: keyof typeof defaultOptions;
} & OptionProps<TAs>;

const DefaultOption = fixedForwardRef<HTMLButtonElement, DefaultOptionProps>(
  (props, ref) => {
    const { kind, ...rest } = props;
    const optionType = defaultOptions[kind] || defaultOptions.help;

    return (
      <Option {...rest} ref={ref}>
        <img src={optionType.src} alt={optionType.alt} />
      </Option>
    );
  },
) as <TAs extends ElementType = 'button'>(
  props: DefaultOptionProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;

const Help = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (props, ref) => {
    return <DefaultOption {...props} kind="help" ref={ref} />;
  },
) as OptionReturnType;

const Close = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (props, ref) => {
    return <DefaultOption {...props} kind="close" ref={ref} />;
  },
) as OptionReturnType;

const Maximize = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (props, ref) => {
    return <DefaultOption {...props} kind="maximize" ref={ref} />;
  },
) as OptionReturnType;

const Minimize = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (props, ref) => {
    return <DefaultOption {...props} kind="minimize" ref={ref} />;
  },
) as OptionReturnType;

const Restore = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>(
  (props, ref) => {
    return <DefaultOption {...props} kind="restore" ref={ref} />;
  },
) as OptionReturnType;

export interface TitleBarBackgroundProps
  extends Omit<HTMLAttributes<HTMLDivElement>, 'color'>,
    FrameProps<'div'> {
  active?: boolean;
  icon?: ReactElement;
  title?: string;
}

interface TitleBarOptions {
  Option: typeof Option;
  OptionsBox: typeof OptionsBox;
  Help: typeof Help;
  Close: typeof Close;
  Maximize: typeof Maximize;
  Minimize: typeof Minimize;
  Restore: typeof Restore;
}

export interface TitleBarProps
  extends TitleBarBackgroundProps,
    TitleBarOptions,
    FrameProps<'div'> {}

const TitleBarRenderer: ForwardRefRenderFunction<
  HTMLDivElement,
  TitleBarBackgroundProps
> = (
  { children, title = 'UNKNOWN.EXE', icon, active = true, ...rest },
  ref: Ref<HTMLDivElement>,
) => (
  <Frame
    {...rest}
    className={cn(styles.titleBarBackground({ active }), rest.className)}
    ref={ref}
  >
    {icon && icon}
    <div className={styles.title}>{title}</div>

    {children}
  </Frame>
);

export const TitleBar = Object.assign(
  fixedForwardRef<HTMLDivElement, TitleBarBackgroundProps>(TitleBarRenderer),
  {
    Option,
    OptionsBox,
    Help,
    Close,
    Maximize,
    Minimize,
    Restore,
  },
) as TitleBarProps & typeof TitleBarRenderer;
