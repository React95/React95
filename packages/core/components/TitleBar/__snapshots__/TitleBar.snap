// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TitleBar /> > Snapshot > should match snapshot 1`] = `
<div>
  <div
    class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683"
  >
    <svg test-file-stub />
    <div
      class="TitleBar_title__1kpzw685"
    >
      test.exe
    </div>
  </div>
  <div
    class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_false__1kpzw684"
  >
    <svg test-file-stub />
    <div
      class="TitleBar_title__1kpzw685"
    >
      test.exe
    </div>
  </div>
  <div
    class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683"
  >
    <div
      class="TitleBar_title__1kpzw685"
    >
      UNKNOWN.EXE
    </div>
  </div>
  <div
    class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_false__1kpzw684"
  >
    <div
      class="TitleBar_title__1kpzw685"
    >
      UNKNOWN.EXE
    </div>
  </div>
</div>
`;

exports[`<TitleBar /> > Snapshot > should match snapshot with all Options 1`] = `
<div>
  <div
    class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683"
  >
    <svg test-file-stub />
    <div
      class="TitleBar_title__1kpzw685"
    >
      test.exe
    </div>
    <div
      class="TitleBar_optionsBox__1kpzw680"
    >
      <button
        class="TitleBar_option__1kpzw681 Button_button__1bzq74l0"
      >
        ?
      </button>
      <button
        class="TitleBar_option__1kpzw681 Button_button__1bzq74l0"
      >
        x
      </button>
    </div>
  </div>
</div>
`;
